src/Language/Pirouette/Example/Syntax.hs

```diff
@@ -112,6 +116,27 @@ tInt, tBool :: Type Ex
 tInt = SystF.tyPure $ SystF.Free $ TyBuiltin TyInteger
 tBool = SystF.tyPure $ SystF.Free $ TyBuiltin TyBool
 
+instance LanguageSMT Ex where
+  translateBuiltinType TyInteger = SimpleSMT.tInt
+  translateBuiltinType TyBool = SimpleSMT.tBool
+  translateConstant (ConstInt n) = SimpleSMT.int n
+  translateConstant (ConstBool b) = SimpleSMT.bool b
+  translateBuiltinTerm TermAdd [x, y] = Just $ SimpleSMT.add x y
+  translateBuiltinTerm TermSub [x, y] = Just $ SimpleSMT.sub x y
+  translateBuiltinTerm TermLt [x, y] = Just $ SimpleSMT.lt x y
+  translateBuiltinTerm TermEq [x, y] = Just $ SimpleSMT.eq x y
+  translateBuiltinTerm TermIte [_, c, t, e] = Just $ SimpleSMT.ite c t e
+  translateBuiltinTerm _ _ = Nothing
+
+instance LanguageSMTBranches Ex where
+  branchesBuiltinTerm TermIte [SystF.TyArg _, SystF.TermArg c, SystF.TermArg t, SystF.TermArg e] =
```

Comment:
This is an interesting technique! I can see how it might be helpful for doing other interesting techniques; like teaching the SMT that `sha256` must be injective, for instance.

Reply:
That's exactly the plan, teaching information aside the result itself (because maybe it's not known).